File: rest-service/api/src/main/java/org/veupathdb/service/multiblast/controller/GuestLinkController.java
package org.veupathdb.service.multiblast.controller;

import javax.ws.rs.ForbiddenException;
import javax.ws.rs.InternalServerErrorException;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.Request;

import mb.lib.db.JobDBManager;
import org.veupathdb.lib.container.jaxrs.providers.UserProvider;
import org.veupathdb.lib.container.jaxrs.server.annotations.Authenticated;
import org.veupathdb.service.multiblast.generated.model.IOGuestLink;
import org.veupathdb.service.multiblast.generated.resources.LinkGuest;

@Authenticated
public class GuestLinkController implements LinkGuest
{
  private final Request req;

  public GuestLinkController(@Context Request req) {
    this.req = req;
  }

  @Override
  public void convGuestToUser(IOGuestLink link) {
    var user = UserProvider.lookupUser(req).orElseThrow(InternalServerErrorException::new);

    try (var db = new JobDBManager()) {
      if (db.userIsGuest(link.getGuestID())) {
        db.updateJobOwner(link.getGuestID(), user.getUserID());
      } else {
        throw new ForbiddenException();
      }
    } catch (Exception e) {
      throw new InternalServerErrorException(e);
    }
  }
}

File: service-query/src/main/kotlin/org/veupathdb/service/mblast/query/controller/GuestLinkController.kt
package org.veupathdb.service.mblast.query.controller

import jakarta.ws.rs.BadRequestException
import jakarta.ws.rs.ForbiddenException
import jakarta.ws.rs.core.Context
import org.glassfish.jersey.server.ContainerRequest
import org.veupathdb.lib.container.jaxrs.server.annotations.Authenticated
import org.veupathdb.service.mblast.query.generated.model.GuestJobTransferRequest
import org.veupathdb.service.mblast.query.generated.resources.LinkGuest
import org.veupathdb.service.mblast.query.sql.JobDBManager

@Authenticated(allowGuests = false)
class GuestLinkController(@Context request: ContainerRequest) : ControllerBase(request), LinkGuest {
  override fun postLinkGuest(entity: GuestJobTransferRequest): LinkGuest.PostLinkGuestResponse {
    if (entity.guestID == null)
      throw BadRequestException()

    when (JobDBManager.userIsGuest(entity.guestID)) {
      null  -> throw ForbiddenException()
      false -> throw ForbiddenException()
      else  -> JobDBManager.updateUserLinksOwner(entity.guestID, userID)
    }

    return LinkGuest.PostLinkGuestResponse.respond204()
  }
}